beluga-mqtt/src/lib.rs

```diff
@@ -240,35 +232,137 @@ impl MqttClient {
+            .into_iter()
+            .map(ToOwned::to_owned)
+            .collect::<Vec<_>>();
+
+        tokio::spawn(async move {
+            manager
+                .lock()
+                .await
+                .schedule_unsubscribe_many(topics.iter().map(AsRef::as_ref));
+        });
+    }
+
+    /// Returns a reference to the name of the thing.
+    pub fn thing_name(&self) -> &str {
+        &self.thing_name
+    }
 }
 
 async fn poll(
+    client: AsyncClient,
     mut event_loop: EventLoop,
-    subscribers: Arc<Mutex<HashMap<String, Sender<Publish>>>>,
+    mut close_rx: mpsc::Receiver<()>,
+    manager: Arc<Mutex<SubscriberManager>>,
 ) {
     loop {
-        match event_loop.poll().await {
-            Ok(event) => {
-                if let Event::Incoming(Packet::Publish(packet)) = event {
-                    let mut subs = subscribers.lock().await;
-                    let topic = packet.topic.clone();
-                    if let Some(subscriber) = subs.get(&topic) {
-                        if let Err(err) = subscriber.send(packet) {
-                            subs.remove(&topic);
-                            error!(error = &err as &dyn std::error::Error, topic = %topic, "couldn't provide packet for a subscriber")
-                        }
-                    }
-                }
+        tokio::select! {
+            event = event_loop.poll() => {
+                process_event(&client, event, manager.lock().await.deref_mut()).await;
             }
-            Err(conn_err) => {
-                error!(
-                    error = &conn_err as &dyn std::error::Error,
-                    "connection error during polling"
-                );
+            _ = close_rx.recv() => {
+                debug!("exit event poll loop");
+                break;
             }
         }
     }
 }
+
+async fn process_event(
+    client: &AsyncClient,
+    event: std::result::Result<Event, ConnectionError>,
+    manager: &mut SubscriberManager,
+) {
+    match event {
+        Ok(event) => {
+            if let Event::Incoming(Packet::Publish(packet)) = event {
+                process_packet(packet, client, manager).await;
+            }
+        }
+        Err(conn_err) => {
+            error!(
+                error = &conn_err as &dyn std::error::Error,
+                "connection error during polling"
+            );
+        }
+    }
+}
+
+async fn process_packet(packet: Publish, client: &AsyncClient, manager: &mut SubscriberManager) {
+    // Unsubscribes the client from the topics that are currently scheduled for
+    // unsubscription.
+    let removed_topics = futures::stream::iter(manager.scheduled())
+        .map(ToOwned::to_owned)
+        .map(|topic| async move {
+            client.unsubscribe(&topic).await?;
+            Ok::<String, rumqttc::ClientError>(topic)
+        })
+        .buffer_unordered(10)
+        .filter_map(|res| async { res.ok() })
+        .collect::<Vec<_>>()
+        .await;
```

Comment:
Does this work right now? I assume it does but something seems off to me.
This `process_packet` function essentially blocks `event_loop.poll()` since it happens as part of the event handling.
Now the problem I'm seeing is that `event_loop.poll()` is what actually drives rumqtt's logic, so how can we unsubscribe here?

Is there something I'm missing?

Reply:
From my debugging it puts messages in a queue, seems like it's completely safe to do these things in a poll.

Reply:
@silvestrpredko I looked into the rumqttc source code. They're using flume for the underlying channel and `unsubscribe` calls [Sender::send_async](https://docs.rs/flume/latest/flume/struct.Sender.html#method.send_async), which will "block" when the channel is full.

Reply:
Yeah, that makes sense.